### PR TITLE
Generic joints

### PR DESCRIPTION
**Pull request by schmidma**
## Introduced Changes

Introduces a generic type parameter to all `XxxJoints` structs. This enables things like `Joints<Duration>` or `Joints<bool>` and will be used in #188 .

## ToDo / Known Issues

none

## Ideas for Next Iterations (Not This PR)

- drop all occurrences of `Joints<f32>` and introduce type safe angle/radian/degree computation.

## How to Test

Everything should still compile as no implementation was changed, only their signatures

### CODE REVIEW

File: crates/types/src/joints.rs [hunk 1]
```diff
@@ -73,57 +64,69 @@ impl Sub for HeadJoints {
...
+    pub shoulder_roll: T,
+    pub elbow_yaw: T,
+    pub elbow_roll: T,
+    pub wrist_yaw: T,
+    pub hand: T,
+}
+
+impl<T> ArmJoints<T>
+where
+    T: Clone,
```
Comment (h3ndrk) [inline crates/types/src/joints.rs:111]: Is `Clone` a `Copy` for `f32`? Or do we have a performance issue here?

File: crates/types/src/joints.rs [hunk 2]
```diff
@@ -231,72 +264,88 @@ impl Sub for LegJoints {
     }
 }
 
-impl AbsDiffEq for LegJoints {
```
Comment (h3ndrk) [inline crates/types/src/joints.rs:234]: Why did you remove that? And also `RelativeEq`?

**Review by h3ndrk — CHANGES_REQUESTED** [review]

File: crates/types/src/joints.rs [hunk 3]
```diff
@@ -4,55 +4,43 @@ use std::{
...
 use serde::{Deserialize, Serialize};
 use serialize_hierarchy::SerializeHierarchy;
 use splines::impl_Interpolate;
 
 #[derive(Clone, Copy, Debug, Default, Deserialize, PartialEq, Serialize, SerializeHierarchy)]
-pub struct HeadJoints {
-    pub yaw: f32,
-    pub pitch: f32,
+#[serialize_hierarchy(bound = "T: SerializeHierarchy + Serialize, for<'de> T: Deserialize<'de>")]
+pub struct HeadJoints<T = f32> {
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/types/src/joints.rs:13]: Yes, I also like the explicit variant more. If you are willing to merge the huuuuge refactoring of all `Joints<f32>` i'm going to do that. I just wanted to keep the changes small. Should I refactor everything?

File: crates/types/src/joints.rs (hunk 1 above)
Comment (schmidma) [reply crates/types/src/joints.rs:111]: ?!?! what should it do differently? There should never be a difference of clone and copy. `Copy` is just a marker trait for the compiler and requires `Clone` to be more strict.

File: crates/types/src/joints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/types/src/joints.rs:13]: Vote for refactoring. I bet it is not that much.

File: crates/types/src/joints.rs (hunk 1 above)
Comment (h3ndrk) [reply crates/types/src/joints.rs:111]: Ok

File: crates/types/src/joints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/types/src/joints.rs]: I did the requested refactoring. All occurrences of `Joints` should be replaced with `Joints<f32>`

[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/types/src/joints.rs]: Thanks

**Review by h3ndrk — APPROVED** [review]
